spec.md

```diff
@@ -1090,6 +1156,346 @@ message ControllerServiceCapability {
+     READY = 2;
+  }
+  Type type = 1;
+}
+```
+
+##### CreateSnapshot Errors
+
+If the plugin is unable to complete the CreateSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a snapshot corresponding to the specified snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
+| Operation pending for snapshot | 10 ABORTED | Indicates that there is a already an operation pending for the specified snapshot. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per snapshot at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same snapshot. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified snapshot, and then retry with exponential back off. |
+| Call not implemented | 12 UNIMPLEMENTED | CreateSnapshot call is not implemented by the plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. Caller MAY call `ControllerGetCapabilities` to discover Plugin capabilities. |
+| Not enough space to create snapshot | 13 RESOURCE_EXHAUSTED | There is not enough space on the storage system to handle the create snapshot request. | Caller should fail this request. Future calls to CreateSnapshot may succeed if space is freed up. |
+
+
+#### `GetSnapshot`
+
+A Controller Plugin MUST implement this RPC call if it has REQUIRES_GET_SNAPSHOT capability.
+This RPC will be called by the CO to get a snapshot.
+GetSnapshot function is synchronous and it should block until the upload is complete and the snapshot is ready.
+GetSnapshot should be called right after CreateSnapshot.
+
+Some cloud providers will upload the snapshot to a location in the cloud (i.e., an object store) after the snapshot is cut.
+Uploading is a long process that could take hours.
+At this phase, since the snapshot is already cut, a `thaw` operation can be performed so application can start running again (assuming a `freeze` operation was done before taking the snapshot).
+For this reason, a GetSnapshot RPC is introduced.
+CreateSnapshot is a synchronous call and it must block until the snapshot is cut.
+GetSnapshot is also synchronous and it must block until the upload process is complete.
```

Comment:
It's still not clear to me why `GetSnapshot` should be blocked until the upload is completed. Does that mean `GetSnapshot` will never know if a snapshot is `Uploading` or not? Even `Uploading` is a valid state in the snapshot status.

If instead GetSnapshot only reports the current status of the snapshot (potentially during uploading), the caller can simply choose to use GetSnapshot to poll the status of the snapshot, probably we can introduce a percentage of progress in the status as well. Blocking for potentially hours is more error-prone IMO.

If so, we may need a Error state in the snapshot object then.

Reply:
OK, just saw the ListSnapshots call with snapshot ID, guess that answer the question.

Though in that case, without an Error state in the snapshot object, it seems hard for the caller to know if one snapshot goes wrong. Or the caller always need to depend on `GetSnapshot` to know if a snapshot goes wrong?  But `GetSnapshot` is optional by itself.

Reply:
GetSnapshot is only required if the storage system needs to do an upload after the snapshot is cut.  That's why it is optional.  Not every storage system do an upload after the snapshot is cut.

Error status is not there because the original thought was that there will be an gRPC error code returned when an error occurs, but I agree that it is better to add the error status there.

Reply:
Added an error status.